README.md
# JSON Schema to OpenAPI Schema

A little NodeJS package to convert JSON Schema to a [OpenAPI Schema Object](http://spec.openapis.org/oas/v3.0.3.html#schema-object).

[![Treeware](https://img.shields.io/badge/dynamic/json?color=brightgreen&label=Treeware&query=%24.total&url=https%3A%2F%2Fpublic.offset.earth%2Fusers%2Ftreeware%2Ftrees)](https://treeware.earth)

## Features

- converts JSON Schema Draft 04 to OpenAPI 3.0 Schema Object
- switches `type: ['foo', 'null']` to `type: foo` and `nullable: true`
- supports deep structures with nested `allOf`s etc.
- switches `patternProperties` to `x-patternProperties`
- converts `dependencies` to an allOf + oneOf OpenAPI-valid equivalent

## Installation

```shell
npm install --save @openapi-contrib/json-schema-to-openapi-schema
```

Requires NodeJS v10 or greater.

## Usage

Here's a small example to get the idea:

```js
const convert = require('@openapi-contrib/json-schema-to-openapi-schema');

const schema = {
	$schema: 'http://json-schema.org/draft-04/schema#',
	type: ['string', 'null'],
	format: 'date-time',
};

(async () => {
	const convertedSchema = await convert(schema);
	console.log(convertedSchema);
})();
```

The example prints out

```js
{
  type: 'string',
  format: 'date-time',
  nullable: true
}
```

### Options

The function accepts `options` object as the second argument.

#### `cloneSchema` (boolean)

If set to `false`, converts the provided schema in place. If `true`, clones the schema by converting it to JSON and back. The overhead of the cloning is usually negligible. Defaults to `true`.

#### `dereference` (boolean)

If set to `true`, all local and remote references (http/https and file) $refs will be dereferenced. Defaults to `false`.

#### `convertUnreferencedDefinitions` (boolean)

Defaults to true.

If a schema had a definitions property (which is valid in JSONSchema), and only some of those entries are referenced, we'll still try and convert the remaining definitions to OpenAPI. If you do not want this behavior, set this to `false`.

#### `dereferenceOptions` (object = $RefParser.Options)

Options to pass to the dereferencer (@apidevtools/json-schema-ref-parser). To prevent circular references, pass `{ dereference: { circular: 'ignore' } }`.

## Command Line

```sh
Usage:
    json-schema-to-openapi-schema <command> [options] <file>

Commands:
    convert                 Converts JSON Schema Draft 04 to OpenAPI 3.0 Schema Object

Options:
    -h, --help              Show help for any command
    -v, --version           Output the CLI version number
    -d, --dereference       If set all local and remote references (http/https and file) $refs will be dereferenced
```

## Why?

OpenAPI is often described as an extension of JSON Schema, but both specs have changed over time and grown independently. OpenAPI v2 was based on JSON Schema draft v4 with a long list of deviations, but OpenAPI v3 shrank that list, upping their support to draft v4 and making the list of discrepancies shorter. This has been solved for OpenAPI v3.1, but for those using OpenAPI v3.0, you can use this tool to solve [the divergence](https://apisyouwonthate.com/blog/openapi-and-json-schema-divergence).

![Diagram showing data model (the objects, payload bodies, etc) and service model (endpoints, headers, metadata, etc)](https://cdn-images-1.medium.com/max/1600/0*hijIL-3Xa5EFZ783.png)

This tool sets out to allow folks to convert from JSON Schema (their one source of truth for everything) to OpenAPI (a thing for HTML docs and making SDKs).

## Versions

- **From:** [JSON Schema Draft v5 †](http://json-schema.org/specification-links.html#draft-5)
- **To:** [OpenAPI 3.0](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md)

_† Draft v5 is also known as Draft Wright 00, as the drafts are often named after the author, and this was the first one by A. Wright. Amongst other things, draft v5 aimed to rewrite the meta files, but the experiment failed, meaning we need to continue to use the draft v4 metafiles. Ugh._

## Converting Back

To convert the other way, check out [openapi-schema-to-json-schema], which this package was based on.

## Tests

To run the test-suite:

```shell
npm test
```

## Treeware

This package is [Treeware](https://treeware.earth). If you use it in production, then we ask that you [**buy the world a tree**](https://plant.treeware.earth/{venfor}/{package}) to thank us for our work. By contributing to the Treeware forest you’ll be creating employment for local families and restoring wildlife habitats.

## Thanks

- [Stoplight][] for [donating time and effort](https://stoplight.io/blog/companies-supporting-open-source/) to this project, and many more.
- [mikunn][] for creating [openapi-schema-to-json-schema] which this is based on.
- [Phil Sturgeon][] for flipping that conversion script about face.
- [WeWork][] for giving this a home for a while.
- [All Contributors][link-contributors]

[mikunn]: https://github.com/mikunn
[wework]: https://github.com/wework
[stoplight]: https://stoplight.io/
[phil sturgeon]: https://github.com/philsturgeon
[openapi-schema-to-json-schema]: https://github.com/openapi-contrib/openapi-schema-to-json-schema
[link-contributors]: https://github.com/openapi-contrib/json-schema-to-openapi-schema/graphs/contributors


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.0.0] - 2020-01-20

### Changed

- Moved over to the `@openapi-contrib` NPM organization.

## [0.4.0] - 2019-10-04

### Added

- Take the first JSON Schema `example` and put in OpenAPI Schema Object `example`

## [0.3.0] - 2018-12-18

### Added

- Create empty items, as it must always be present for type: array
- Rewrite exclusiveMinimum/exclusiveMaximum
- Rewrite if/then/else as oneOf + allOf
- Rewrite const as single element enum

## [0.2.0] - 2018-05-10

### Fixed

- Implemented [@cloudflare/json-schema-walker] to make sure all subschemas are
  processed

[@cloudflare/json-schema-walker]: https://github.com/cloudflare/json-schema-tools#cloudflarejson-schema-walker

## [0.1.1] - 2018-04-09

### Added

- Convert `dependencies` to an allOf + oneOf OpenAPI-valid equivalent


test/items.js
'use strict';

const convert = require('../');
const should = require('should');

it('items', () => {
	const schema = {
		$schema: 'http://json-schema.org/draft-04/schema#',
		type: 'array',
		items: {
			type: 'string',
			format: 'date-time',
			example: '2017-01-01T12:34:56Z'
		}
	};

	const result = convert(schema);

	const expected = {
		type: 'array',
		items: {
			type: 'string',
			format: 'date-time',
			example: '2017-01-01T12:34:56Z'
		}
	};

	should(result).deepEqual(expected, 'converted');
});


test/helpers.js
const fs = require('fs');
const join = require('path').join;

const getSchema = file => {
	const path = join(__dirname, 'schemas', file);
	return JSON.parse(fs.readFileSync(path));
};

module.exports = { getSchema };


test/helpers.ts
import fs from 'fs';
import { join } from 'path';

export const getSchema = (file: string) => {
	const path = join(__dirname, 'schemas', file);
	return JSON.parse(fs.readFileSync(path).toString());
};


index.js
'use strict';

const schemaWalker = require('@cloudflare/json-schema-walker');
const { Resolver } = require('@stoplight/json-ref-resolver');
const { parse } = require('@stoplight/yaml');
const fetch = require('node-fetch');
const oas3schema = require('./refs/oas3-schema.json');

class InvalidTypeError extends Error {
  constructor(message) {
    super()
    this.name = 'InvalidTypeError';
    this.message = message;
  }
}

class DereferencingError extends Error {
  constructor(errors) {
    super()
    this.name = 'DereferencingError';
    this.errors = errors;
  }
}

async function convert(schema, options = {}) {
	const { cloneSchema = true, dereference = false } = options;

	if (cloneSchema) {
		schema = JSON.parse(JSON.stringify(schema));
	}

	if (dereference) {
		const result = await resolver.resolve(schema);
    if (result.errors && result.errors.length > 0) {
      throw new DereferencingError(result.errors)
    }
    schema = result.result;
	}

	const vocab = schemaWalker.getVocabulary(schema, schemaWalker.vocabularies.DRAFT_04);
	schemaWalker.schemaWalk(schema, convertSchema, null, vocab);
	return schema;
}

function stripIllegalKeywords(schema) {
	delete schema['$schema'];
	delete schema['$id'];
	delete schema['id'];
	return schema;
}

function convertSchema(schema, path, parent, parentPath) {
	schema = stripIllegalKeywords(schema);
	schema = convertTypes(schema);
	schema = rewriteConst(schema);
	schema = convertDependencies(schema);
	schema = rewriteIfThenElse(schema);
	schema = rewriteExclusiveMinMax(schema);
	schema = convertExamples(schema);

	if (typeof schema['patternProperties'] === 'object') {
		schema = convertPatternProperties(schema);
	}

	if (schema.type === 'array' && typeof schema.items === 'undefined') {
		schema.items = {};
  }

  // should be called last
  schema = convertIllegalKeywordsAsExtensions(schema);
	return schema;
}

function validateType(type) {
	const validTypes = ['null', 'boolean', 'object', 'array', 'number', 'string', 'integer'];
	const types = Array.isArray(type) ? type : [type];
	types.forEach(type => {
		if (validTypes.indexOf(type) < 0 && type !== undefined)
			throw new InvalidTypeError('Type "' + type + '" is not a valid type');
	});
}

function convertDependencies(schema) {
	const deps = schema.dependencies;
	if (typeof deps !== 'object') {
		return schema;
	}

	// Turns the dependencies keyword into an allOf of oneOf's
	// "dependencies": {
	// 		"post-office-box": ["street-address"]
	// },
	//
	// becomes
	//
	// "allOf": [
	// 	{
	// 		"oneOf": [
	// 			{"not": {"required": ["post-office-box"]}},
	// 			{"required": ["post-office-box", "street-address"]}
	// 		]
	// 	}
	//

	delete schema['dependencies'];
	if (!Array.isArray(schema.allOf)) {
		schema.allOf = [];
	}

	for (const key in deps) {
		const foo = {
			'oneOf': [
				{
					'not': {
						'required': [key]
					}
				},
				{
					'required': [].concat(key, deps[key])
				}
			]
		};
		schema.allOf.push(foo);
	}
	return schema;
}

function convertTypes(schema) {
	if (schema.type === undefined) {
		return schema;
	}

	validateType(schema.type);

	if (Array.isArray(schema.type)) {
		if (schema.type.includes('null')) {
			schema.nullable = true;
		}
		const typesWithoutNull = schema.type.filter(type => type !== 'null');
		if (typesWithoutNull.length === 0) {
			delete schema.type
		} else if (typesWithoutNull.length === 1) {
			schema.type = typesWithoutNull[0];
		} else {
			delete schema.type;
			schema.anyOf = typesWithoutNull.map(type => ({ type }));
		}
	}
	else if (schema.type === 'null') {
		delete schema.type;
		schema.nullable = true;
	}

	return schema;
}

// "patternProperties did not make it into OpenAPI v3.0"
// https://github.com/OAI/OpenAPI-Specification/issues/687
function convertPatternProperties(schema) {
	schema['x-patternProperties'] = schema['patternProperties'];
	delete schema['patternProperties'];
	if (typeof schema.additionalProperties === 'undefined') schema.additionalProperties = true;
	return schema;
}

// keywords (or property names) that are not recognized within OAS3 are rewritten into extensions.
function convertIllegalKeywordsAsExtensions(schema) {
  Object.keys(schema)
    .filter(keyword => !keyword.startsWith(oasExtensionPrefix) && !allowedKeywords.includes(keyword))
    .forEach(keyword => {
      schema[oasExtensionPrefix + keyword] = schema[keyword];
      delete schema[keyword];
    });
  return schema;
}

function convertExamples(schema) {
  if (schema['examples'] && Array.isArray(schema['examples'])) {
    schema['example'] = schema['examples'][0];
    delete schema['examples'];
  }

  return schema;
}

function rewriteConst(schema) {
	if (Object.hasOwnProperty.call(schema, 'const')) {
		schema.enum = [ schema.const ];
		delete schema.const;
	}
	return schema;
}

function rewriteIfThenElse(schema) {
/* @handrews https://github.com/OAI/OpenAPI-Specification/pull/1766#issuecomment-442652805
if and the *Of keywords

There is a really easy solution for implementations, which is that

if: X, then: Y, else: Z

is equivalent to

oneOf: [allOf: [X, Y], allOf: [not: X, Z]]
*/
	if (schema.if && schema.then) {
		schema.oneOf = [ { allOf: [ schema.if, schema.then ] },
				 { allOf: [ { not: schema.if }, schema.else ] } ];
		delete schema.if;
		delete schema.then;
		delete schema.else;
  }
	return schema;
}

function rewriteExclusiveMinMax(schema) {
	if (typeof schema.exclusiveMaximum === 'number') {
		schema.maximum = schema.exclusiveMaximum;
		schema.exclusiveMaximum = true;
	}
	if (typeof schema.exclusiveMinimum === 'number') {
		schema.minimum = schema.exclusiveMinimum;
		schema.exclusiveMinimum = true;
	}
	return schema;
}

const httpReader = {
	async resolve(ref) {
		return (await fetch(String(ref))).text();
	},
};

const resolver = new Resolver({
	resolvers: {
		http: httpReader,
		https: httpReader,
		file: {
			resolve: async (ref) => {
				const fs = require('fs');
				const readFileAsync = (fs.promises && fs.promises.readFile) || require('util').promisify(fs.readFile);
				return readFileAsync(ref.path(), 'utf8');
			},
		},
	},

	transformDereferenceResult(opts) {
		// this should not be needed, but unfortunately is due to a quirk in json-ref-resolver returning sealed object
		opts.result = JSON.parse(JSON.stringify(opts.result));
		return opts;
	},

	parseResolveResult(opts) {
		try {
			opts.result = parse(opts.result);
		} catch {
			// let's carry on
		}

		return opts;
	},
});

const oasExtensionPrefix = 'x-';

// TODO: having definitions inside an oas3 schema isn't exactly valid,
// maybe it is an idea to extract and split them into multiple oas3 schemas and reference to them.
// For now leaving as is.
const allowedKeywords = ['$ref', 'definitions'].concat(Object.keys(oas3schema.definitions.Schema.properties));

module.exports = convert;


.eslintrc.js
module.exports = {
	env: {
		browser: true,
		commonjs: true,
		es6: true,
		node: true,
	},
	extends: [
		'prettier',
		'eslint:recommended',
		'plugin:@typescript-eslint/recommended',
	],
	parser: '@typescript-eslint/parser',
	globals: {
		Atomics: 'readonly',
		SharedArrayBuffer: 'readonly',
	},
	parserOptions: {
		ecmaVersion: 2018,
	},
	plugins: ['prettier', 'unused-imports', '@typescript-eslint'],
	rules: {
		indent: [
			'error',
			'tab',
			{
				SwitchCase: 1,
			},
		],
		'linebreak-style': [
			'error',
			process.platform === 'win32' ? 'windows' : 'unix',
		],
		quotes: ['error', 'single'],
		semi: ['error', 'always'],
		'@typescript-eslint/ban-ts-comment': 'off',
		'@typescript-eslint/consistent-type-definitions': ['error', 'interface'],
		'@typescript-eslint/consistent-type-imports': [
			'error',
			{
				prefer: 'type-imports',
			},
		],
		'@typescript-eslint/no-explicit-any': 'off',
	},
	ignorePatterns: ['dist/**', 'bin/**'],
};


src/index.ts
import type { JSONSchema } from '@apidevtools/json-schema-ref-parser';
import type {
	JSONSchema4,
	JSONSchema6Definition,
	JSONSchema7Definition,
} from 'json-schema';
import type { Options, SchemaType, SchemaTypeKeys } from './types';
import { Walker } from 'json-schema-walker';
import { allowedKeywords } from './const.js';
import type { OpenAPIV3 } from 'openapi-types';

class InvalidTypeError extends Error {
	constructor(message: string) {
		super(message);
		this.name = 'InvalidTypeError';
		this.message = message;
	}
}

const oasExtensionPrefix = 'x-';

const handleDefinition = async <T>(
	def: JSONSchema7Definition | JSONSchema6Definition | JSONSchema4,
	schema: T
) => {
	if (typeof def !== 'object') {
		return def;
	}

	const type = def.type;
	if (type) {
		// Walk just the definitions types
		const walker = new Walker<T>();
		await walker.loadSchema(
			{
				definitions: schema['definitions'] || [],
				...def,
				$schema: schema['$schema'],
			} as any,
			{
				dereference: true,
				cloneSchema: true,
				dereferenceOptions: {
					dereference: {
						circular: 'ignore',
					},
				},
			}
		);
		await walker.walk(convertSchema, walker.vocabularies.DRAFT_07);
		if ('definitions' in walker.rootSchema) {
			delete (<any>walker.rootSchema).definitions;
		}
		return walker.rootSchema;
	} else if (Array.isArray(def)) {
		// if it's an array, we might want to reconstruct the type;
		const typeArr = def;
		const hasNull = typeArr.includes('null');
		if (hasNull) {
			const actualTypes = typeArr.filter((l) => l !== 'null');
			return {
				type: actualTypes.length === 1 ? actualTypes[0] : actualTypes,
				nullable: true,
				// this is incorrect but thats ok, we are in the inbetween phase here
			} as JSONSchema7Definition | JSONSchema6Definition | JSONSchema4;
		}
	}

	return def;
};

const convert = async <T = JSONSchema>(
	schema: T,
	options?: Options
): Promise<OpenAPIV3.Document> => {
	const walker = new Walker<T>();
	const convertDefs = options?.convertUnreferencedDefinitions ?? true;
	await walker.loadSchema(schema, options);
	await walker.walk(convertSchema, walker.vocabularies.DRAFT_07);
	// if we want to convert unreferenced definitions, we need to do it iteratively here
	const rootSchema = walker.rootSchema as unknown as JSONSchema;
	if (convertDefs && rootSchema?.definitions) {
		for (const defName in rootSchema.definitions) {
			const def = rootSchema.definitions[defName];
			rootSchema.definitions[defName] = await handleDefinition(def, schema);
		}
	}
	return rootSchema as OpenAPIV3.Document;
};

function stripIllegalKeywords(schema: SchemaType) {
	if (typeof schema !== 'object') {
		return schema;
	}
	delete schema['$schema'];
	delete schema['$id'];
	if ('id' in schema) {
		delete schema['id'];
	}
	return schema;
}

function convertSchema(schema: SchemaType | undefined) {
	if (!schema) {
		return schema;
	}
	schema = stripIllegalKeywords(schema);
	schema = convertTypes(schema);
	schema = rewriteConst(schema);
	schema = convertDependencies(schema);
	schema = rewriteIfThenElse(schema);
	schema = rewriteExclusiveMinMax(schema);
	schema = convertExamples(schema);

	if (typeof schema['patternProperties'] === 'object') {
		schema = convertPatternProperties(schema);
	}

	if (schema.type === 'array' && typeof schema.items === 'undefined') {
		schema.items = {};
	}

	// should be called last
	schema = convertIllegalKeywordsAsExtensions(schema);
	return schema;
}
const validTypes = new Set([
	'null',
	'boolean',
	'object',
	'array',
	'number',
	'string',
	'integer',
]);
function validateType(type: any) {
	if (typeof type === 'object' && !Array.isArray(type)) {
		// Refs are allowed because they fix circular references
		if (type.$ref) {
			return;
		}
		// this is a de-referenced circular ref
		if (type.properties) {
			return;
		}
	}
	const types = Array.isArray(type) ? type : [type];
	types.forEach((type) => {
		if (type && !validTypes.has(type))
			throw new InvalidTypeError('Type "' + type + '" is not a valid type');
	});
}

function convertDependencies(schema: SchemaType) {
	const deps = schema.dependencies;
	if (typeof deps !== 'object') {
		return schema;
	}

	// Turns the dependencies keyword into an allOf of oneOf's
	// "dependencies": {
	// 		"post-office-box": ["street-address"]
	// },
	//
	// becomes
	//
	// "allOf": [
	// 	{
	// 		"oneOf": [
	// 			{"not": {"required": ["post-office-box"]}},
	// 			{"required": ["post-office-box", "street-address"]}
	// 		]
	// 	}
	//

	delete schema['dependencies'];
	if (!Array.isArray(schema.allOf)) {
		schema.allOf = [];
	}

	for (const key in deps) {
		const foo: (JSONSchema4 & JSONSchema6Definition) & JSONSchema7Definition = {
			oneOf: [
				{
					not: {
						required: [key],
					},
				},
				{
					required: [key, deps[key]].flat() as string[],
				},
			],
		};
		schema.allOf.push(foo);
	}
	return schema;
}

function convertTypes(schema: SchemaType) {
	if (typeof schema !== 'object') {
		return schema;
	}
	if (schema.type === undefined) {
		return schema;
	}

	validateType(schema.type);

	if (Array.isArray(schema.type)) {
		if (schema.type.includes('null')) {
			schema.nullable = true;
		}
		const typesWithoutNull = schema.type.filter((type) => type !== 'null');
		if (typesWithoutNull.length === 0) {
			delete schema.type;
		} else if (typesWithoutNull.length === 1) {
			schema.type = typesWithoutNull[0];
		} else {
			delete schema.type;
			schema.anyOf = typesWithoutNull.map((type) => ({ type }));
		}
	} else if (schema.type === 'null') {
		delete schema.type;
		schema.nullable = true;
	}

	return schema;
}

// "patternProperties did not make it into OpenAPI v3.0"
// https://github.com/OAI/OpenAPI-Specification/issues/687
function convertPatternProperties(schema: SchemaType) {
	schema['x-patternProperties'] = schema['patternProperties'];
	delete schema['patternProperties'];
	schema.additionalProperties ??= true;
	return schema;
}

// keywords (or property names) that are not recognized within OAS3 are rewritten into extensions.
function convertIllegalKeywordsAsExtensions(schema: SchemaType) {
	const keys = Object.keys(schema) as SchemaTypeKeys[];
	keys
		.filter(
			(keyword) =>
				!keyword.startsWith(oasExtensionPrefix) &&
				!allowedKeywords.includes(keyword)
		)
		.forEach((keyword: SchemaTypeKeys) => {
			const key = `${oasExtensionPrefix}${keyword}` as keyof SchemaType;
			schema[key] = schema[keyword];
			delete schema[keyword];
		});
	return schema;
}

function convertExamples(schema: SchemaType) {
	if (schema['examples'] && Array.isArray(schema['examples'])) {
		schema['example'] = schema['examples'][0];
		delete schema['examples'];
	}

	return schema;
}

function rewriteConst(schema: SchemaType) {
	if (Object.hasOwnProperty.call(schema, 'const')) {
		schema.enum = [schema.const];
		delete schema.const;
	}
	return schema;
}

function rewriteIfThenElse(schema: SchemaType) {
	if (typeof schema !== 'object') {
		return schema;
	}
	/* @handrews https://github.com/OAI/OpenAPI-Specification/pull/1766#issuecomment-442652805
  if and the *Of keywords

  There is a really easy solution for implementations, which is that

  if: X, then: Y, else: Z

  is equivalent to

  oneOf: [allOf: [X, Y], allOf: [not: X, Z]]
  */
	if ('if' in schema && schema.if && schema.then) {
		schema.oneOf = [
			{ allOf: [schema.if, schema.then].filter(Boolean) },
			{ allOf: [{ not: schema.if }, schema.else].filter(Boolean) },
		];
		delete schema.if;
		delete schema.then;
		delete schema.else;
	}
	return schema;
}

function rewriteExclusiveMinMax(schema: SchemaType) {
	if (typeof schema.exclusiveMaximum === 'number') {
		schema.maximum = schema.exclusiveMaximum;
		schema.exclusiveMaximum = true;
	}
	if (typeof schema.exclusiveMinimum === 'number') {
		schema.minimum = schema.exclusiveMinimum;
		schema.exclusiveMinimum = true;
	}
	return schema;
}

export default convert;
